### PR TITLE
Adding getSupportedPlatformInfoFunctions

### PR DESCRIPTION
**Pull request by schie**
## Description

re #1085

- Added a helper/convenience function, `getSupportedPlatformInfoFunctions(...)` that gets rid of some of the repetitive logic around supported platforms and memoization
- Added helper functions, `getSupportedPlatformInfoSync` and `getSupportedPlatformInfoAsync` , that can be used to explicitly get info sync or async if one desires to do so

## Compatibility

| OS      | Implemented |
| ------- | :---------: |
| iOS     |    ✅     |
| Android |    ✅     |
| Windows |    ✅     |

## Checklist

* [X] I have tested this on a device/simulator for each compatible OS
  - unable to verify on Windows, but did verify for iOS and Android
* [ ] I added the documentation in `README.md`
* [ ] I mentioned this change in `CHANGELOG.md`
* [X] I updated the typings files (`privateTypes.ts`, `types.ts`)
* [ ] I added a sample use of the API (`example/App.js`)

### Screen Shots
Below are screen shots of iOS and Android and the updated `getDeviceName*` functions (no difference)
image

### CODE REVIEW

**Comment by schie** [comment]
The usage of this function can probably be used with a lookup table to make the index file more terse, but that's a future thing if desired — me thinks.

File: src/index.ts [hunk 1]
```diff
@@ -440,28 +440,13 @@ export function getBootloaderSync() {
...
-      device = RNDeviceInfo.getDeviceSync();
-    } else {
-      device = 'unknown';
-    }
-  }
-  return device;
-}
+export const [getDevice, getDeviceSync] = getSupportedPlatformInfoFunctions({
+  getter: () => RNDeviceInfo.getDeviceName(),
+  syncGetter: () => RNDeviceInfo.getDeviceNameSync(),
```
Comment (mikehardy) [inline src/index.ts:445]: I think you intend these native getters for this one?

```suggestion
  getter: () => RNDeviceInfo.getDevice(),
  syncGetter: () => RNDeviceInfo.getDeviceSync(),
```

**Review by mikehardy — CHANGES_REQUESTED** [review]
This looks excellent to me - just one comment that appears to be a trivial think-o

File: src/index.ts (hunk 1 above)
Comment (schie) [reply src/index.ts:445]: good catch!

**Review by mikehardy — APPROVED** [review]
Excellent! I'll merge this but as it's just a refactor and not even a bugfix I'll not do a release, just FYI (no user benefit, seems like a net negative) but this will be a big improvement for development.

If it works for you and you're interested in doing it, normalizing the rest of the file like this would work fine by me :-) :pray:

**Comment by schie** [comment]
Totally fine with me.

I'll knock out a few conversions later today.

**Pull request merged**